.github/workflows/release.yml

```diff
@@ -1,20 +1,49 @@
 name: Release
 
 on:
+  workflow_dispatch:
   release:
     types:
       - published
 
 jobs:
+  dist:
+    runs-on: ubuntu-latest
+    permissions:
+      id-token: write
+      contents: read
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Build SDist and wheel
+        run: pipx run build
+
+      - name: Check metadata
+        run: pipx run twine check dist/*
+
+      - uses: actions/upload-artifact@v4
+        with:
+          path: dist/*
+
   publish:
-    name: Publish on PyPI
+    needs: [dist]
     runs-on: ubuntu-latest
+    if: github.event_name == 'release' && github.event.action == 'published'
     environment:
       name: publish
       url: https://pypi.org/p/glass
     permissions:
       id-token: write
+
     steps:
-      - uses: actions/checkout@v4
```

Comment:
I'm not sure how `checkout` without setting `fetch-depth: 0` was working (or maybe it was silently failing)? Given that glass uses VCS for versioning, it becomes important to pull in the entire git history (not a bad thing) to make the versions/tags work.

Reply:
It should work since it's the release tag itself that is being cloned to a depth of 1 here:

```console
% git clone --branch v2024.1 --depth 1 https://github.com/glass-dev/glass.git glass-v2024.1
% cd glass-v2024.1 
% git describe --tags
v2024.1
```

Reply:
Ah, I see, the workflow is triggered when a release is published so the it `checkout`s on the tag and not the branch. That makes sense, thanks!